core/src/main/scala/config/decoders.scala

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright 2017 47 Degrees, LLC. <http://www.47deg.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package freestyle.cassandra
+package config
+
+import java.util.concurrent.Executor
+
+import classy.{Decoder, Read}
+import classy.config._
+import com.datastax.driver.core.ProtocolOptions.Compression
+import com.datastax.driver.core._
+import com.datastax.driver.core.policies._
+import com.typesafe.config.Config
+
+import collection.JavaConverters._
+
+object decoders {
+
+  def builderCustomProp[Builder, T](b: Builder, attr: String)(f: (Builder, T) => Builder)(
+      implicit r: Read[Config, T]): ConfigDecoder[Builder] =
+    builderProp[Builder, T](b, attr, b => f(b, _))
+
+  def builderProp[Builder, T](b: Builder, attr: String, f: Builder => (T => Builder))(
+      implicit r: Read[Config, T]): ConfigDecoder[Builder] =
+    readConfig[Option[T]](attr).map {
+      case Some(v) => f(b)(v)
+      case _       => b
+    }
+
+  import config.implicits._
+  import config.implicits.datastax._
+
+  abstract class DecoderBuilder[Builder] {
+
+    def empty: Builder
+
+    def fields: List[(Builder) => ConfigDecoder[Builder]]
+
+    def customField[T](b: Builder, attr: String)(f: (Builder, T) => Builder)(
+        implicit r: Read[Config, T]): ConfigDecoder[Builder] =
+      builderCustomProp[Builder, T](b, attr)(f)
+
+    def field[T](b: Builder, attr: String, f: Builder => (T => Builder))(
+        implicit r: Read[Config, T]): ConfigDecoder[Builder] = builderProp[Builder, T](b, attr, f)
+
+    def flagField(b: Builder, attr: String, f: Builder => Builder): ConfigDecoder[Builder] =
+      customField[Boolean](b, attr) {
+        case (builder, true) => f(builder)
+        case (builder, _)    => builder
+      }
+
+    def build: ConfigDecoder[Builder] =
+      fields.foldLeft[ConfigDecoder[Builder]](Decoder.const(empty)) { (d, b) =>
+        d.flatMap(b)
+      }
```

Comment:
Just for fun :smiley:, some different ways to write this:

With a  `classy-cats` dependency:

- we could use `foldM`:

```scala
fields.foldM(empty)((b, d) => d(b))
```

- or `Kleisli`'s `MonoidK` : 

```scala
def fields: List[Kleisli[ConfigDecoder, Builder, Builder]]

def build: ConfigDecoder[Builder] = fields.foldK.run(empty)
```